### ISSUE TITLE
[Feature Request] Build Telegram for Windows on ARM64

### ISSUE THREAD

**Issue by ssg** [telegramdesktop/tdesktop#25266]
### Is your feature request related to a problem?

Telegram runs under x64 emulation on Windows ARM devices, that causes a performance penalty especially with the CPU intensive tasks like video/GIFs, etc. With the introduction of Snapdragon X series processors, ARM64 laptops are now more prevalent.

### Describe the solution you'd like

ARM64 Builds for Windows.

### Describe alternatives you've considered

I've considered suffering through performance problems, I didn't like it.

**Comment by sungaila** [comment]
Telegram cannot be ported to ARM64 just yet because some third-party dependencies are not ready (e.g. Qt ARM64 builds are still in tech preview). But ARM64EC might be a good compromise for now: Telegram runs natively and everything else stays x64.

Also porting can be done iteratively: replace the x64 dependencies with ARM64 builds step by step. Once all third-party dependencies are updated, just change the build target from ARM64EC to ARM64.

**Comment by NicolasOchoaMSFT** [comment]
> Telegram cannot be ported to ARM64 just yet because some third-party dependencies are not ready (e.g. Qt ARM64 builds are still in tech preview). But ARM64EC might be a good compromise for now: Telegram runs natively and everything else stays x64.
> 
> ~Also porting can be done iteratively: replace the x64 dependencies with ARM64 builds step by step. Once all third-party dependencies are updated, just change the build target from ARM64EC to ARM64.~
> 
> Edit: Forget the last part. The Microsoft docs state that ARM64EC can load x64 and ARM64EC binaries but **they cannot load ARM64 binaries**.

Hey Team, 

​​​​​​​My name is Nicolas Ochoa, and I am an App Assure Manager with Microsoft's FastTrack Center. My team assists Microsoft customers and Independent Software Vendors (ISVs) with any application or software compatibility scenarios encountered while upgrading to our latest versions of Microsoft products. 

We would like to work with Telegram to prepare your applications for Windows on Arm.

The App Assure team supports application compatibility for Windows 10, Windows 11, and Windows on Arm with an emphasis on collaborating with ISVs. Our goal is to help ensure your Telegram application is  supported on Windows 11 and Windows on Arm. Our App Assure engineers can provide you with best practice guidelines and help remove any technical blockers encountered when developing compatible applications. These services are provided at no additional cost.

Feel free to reach out to me by replying to this thread

**Comment by ChrisOfTheLewi** [comment]
Arm64EC supports ARM64 code, but not ARM64 binaries.  It works very well for partial ports in conjunction with a good linking strategy (the arm64 codebase in ARM64 and Arm64EC can be the same code essentially and ultimately you can produce Arm64X if needed to support both types from a single binary should you have any dlls which need to be able to load in both types of processes).  Arm64EC is usually where I'd suggest open source projects with lots of dependencies start looking; you can get pretty great performance without having to recompile everything into Arm64.

**Comment by khmyznikov** [comment]
Yes please, definitely need this.

**Comment by ssg** [comment]
> Arm64EC supports ARM64 code, but not ARM64 binaries

Is that a blocker for Telegram?

**Comment by CryptoManiac** [comment]
> Telegram cannot be ported to ARM64 just yet because some third-party dependencies are not ready (e.g. Qt ARM64 builds are still in tech preview). But ARM64EC might be a good compromise for now: Telegram runs natively and everything else stays x64.

QT builds for arm64 without any issues, I checked that more than two years ago. The full source tree including the IDE, no problem at all. Depending on the prebuilt libraries when a source is available seems quite weird for an open source project.

**Comment by sungaila** [comment]
@CryptoManiac Currently the Qt arm64 builds are considered stable but the tooling isn't there yet. See https://www.qt.io/blog/qt-for-windows-on-arm

> Qt basically works natively on Windows on ARM. It is usable to build and run native Qt applications for your ARM64 devices _but you will have to build Qt yourself_. We, the Qt Company, still have some work to do to make the developer experience come up to par with the rest of our offering though.

There are no official binaries yet and I don't think this project wants to built Qt itself.

**Comment by john-preston** [comment]
Well, really this project (if you're talking about TDesktop) builds (and patches) Qt itself on all platforms for ten years now. But I'm not sure about ARM build of all other dependencies, Qt is just one, on Windows there are 28 steps in dependencies-building script. Right now I don't have time to check/port building all of them to ARM and then building one more version for each release :(

**Comment by sungaila** [comment]
>But I'm not sure about ARM build of all other dependencies

I mean, that is the point of ARM64EC: use this target for the core of TDesktop and the (third-party) dependencies can stay x86-64. Then migrate the rest iteratively one step at a time.

>Right now I don't have time to check/port building all of them to ARM and then building one more version for each release :(

I would help out if I had any C++ building experience. I'm still hoping someone else will be interested in giving it a shot. 😅

**Comment by nycevan** [comment]
In light of all the new windows ARM hardware released with the snapdragon chips, I would love to see a native version of telegram.

**Comment by Aokromes** [comment]
i am sure windows arm users is far lower than linux arm users with is far lower than linux x86 users and preston have far higher priorities, when windows arm can run windows x86_64 binaries.

**Comment by grandrew** [comment]
As an active Windows ARM + Telegram user, I prefer the web version because ARM experience of Telegram in WIndows is awful:

1. The UI elements are painfully slow. While app and calling technically does work, pressing every button is about a 5-second delay. And you have to press 3 buttons to make a call
2. It requires 500+ MB of RAM to just sit there while web version is less than 300. Calling on web version of telegram does not work on Edge/ARM. I believe x86 version of telegram is less than 100MB
3. Launching telegram is also about a 5-sec delay and depends on luck

On the overall Windows/ARM projections I believe this will get a lot of traction because it works at least as stable as x86 with more than acceptable performance and provides unbeatable benefits in battery, 5G connectivity, temperature and noise

**Comment by ssg** [comment]
> i am sure windows arm users is far lower than linux arm users wicth is far lower than linux x86 users and preston have far higher priorities, when windows arm can run windows x86_64 binaries.

What's the minimum number of people affected to address an issue?

**Comment by CryptoManiac** [comment]
While the native WoA build is unavailable, you could use Android version. It works fine with Windows Subsystem for Android. It's an ugly crutch of course, but Android version far more usable than Windows x86_64 version. If anything, you'll be able to make calls without spending five minutes on establishing the connection. Hope it helps.

**Comment by CryptoManiac** [comment]
> i am sure windows arm users is far lower than linux arm users wicth is far lower than linux x86 users and preston have far higher priorities, when windows arm can run windows x86_64 binaries.

It works fine as long as you won't try to make a voice or video call. Something is severely broken in the cryptographic keys negotiation code. Broken so much that I can't make a tea before the connection will be established. I guess this can be worked around by using the unoptimized code for this particular use case but this will require modifications.

**Comment by ssg** [comment]
> It works fine as long as you won't try to make a voice or video call.

Nope, autoplay GIFs and videos are problematic too. Telegram is still the top battery consumer on my Surface Laptop 7 without making a voice or video call.

**Comment by ssg** [comment]
> Didn't notice any trouble with them.

Yeah, GIFs and videos play smooth, but they consume a lot of power due to emulation running behind the scenes. That's what I meant as problematic.

**Comment by sungaila** [comment]
I have a Samsung Galaxy Book S (Snapdragon 8cx) and can't confirm any of the mentioned performance issues. Are you sure that you are running the x64 build (NOT x86)?

The battery usage is very poor though judging by the Windows logs.

**Comment by john-preston** [comment]
@ssg Well, from Unigram statistics the ratio of x64 users to arm users is 16'945, meaning there is almost 17'000 people using x64 version on every person using arm version, or arm userbase is ~0.006%

Given that I've EOLed support for systems under 1% of userbase when the time came it's really hard to justify spending big amount of time (working week? idk) to support native arm version for 0.006% of userbase.

BTW for me this percent would be even less, because Unigram works only on modern systems, while I still have relatively big percent of users on Windows 7 (more than 4% of userbase).

idk.. I'm open for the thing, but it is really hard to make (and support, and build every release) a native arm version and really low part of userbase

**Comment by sungaila** [comment]
@john-preston You would have to copy the Windows x64 pipeline, change the build target to Arm64EC and leave everything else (like x64 dependencies) as is. So that's 1-2 days max? But without proper hardware or VMs there is just no way to test it yourself.

Maybe Microsoft (@NicolasOchoaMSFT) would be willing to send you a Windows Dev Kit 2023? 😄

**Comment by john-preston** [comment]
@sungaila I'm not sure building tdesktop code Arm native and leaving dependencies x64 (like OpenSSL / FFmpeg / WebRTC) will do any good in terms of performance, I doubt that tdesktop code is the main problem here. It looks like a fully native build is needed for all this to make sense.

**Comment by john-preston** [comment]
I hope I'll be able to use Mac hardware for Windows on Arm development, that way I still need only two hardware sets to build all of supported tdesktop versions (Mac on Apple Silicon builds both x64 and Arm macOS versions, Windows on x64 builds Windows x86, Windows x64 and Linux x64 versions).

**Comment by CryptoManiac** [comment]
> @john-preston You would have to copy the Windows x64 pipeline, change the build target to Arm64EC and leave everything else (like x64 dependencies) as is. So that's 1-2 days max? But without proper hardware or VMs there is just no way to test it yourself.
> 
> 
> 
> Maybe Microsoft (@NicolasOchoaMSFT) would be willing to send you a Windows Dev Kit 2023? 😄

Raspberry Pi will do. There is no need for expensive hardware to run these binaries.

**Comment by john-preston** [comment]
I don't want another distinct hardware, I really do hope Apple Silicon would do. I already need it for macOS builds.

**Comment by sungaila** [comment]
@john-preston You could be right, I too doubt that tdesktop is the bottleneck here.

For development you can run Windows 11 on ARM via virtualization on your Mac (e.g. https://kb.parallels.com/de/125375).

Windows cross-compiles to x86, x64, arm64 and arm64ec on each platform. You'd need Windows on ARM to test the binaries but not for compiling.

**Comment by CryptoManiac** [comment]
> @john-preston You could be right, I too doubt that tdesktop is the bottleneck here.
> 
> 
> 
> For development you can run Windows 11 on ARM via virtualization on your Mac (e.g. https://kb.parallels.com/en/125375).
> 
> 
> 
> Windows cross-compiles to x86, x64, arm64 and arm64ec on each platform. You'd need Windows on ARM to test the binaries but not for compiling.

UTM works reasonably well. There is no GPU acceleration on Windows yet, but it's still more than enough for testing and development. 

https://mac.getutm.app

**Comment by john-preston** [comment]
I have Windows on Arm in Parallels Studio there. I just didn't manage to go further than first dependency out of 28 that I need to compile in order to build tdesktop on Arm 😢 

The cross-compiling also sucks for tdesktop because there are several code generator built and used (!) before tdesktop itself, so the resulting process is very complicated if you try to cross-compile. You need to build some libraries and codegens in host arch, then make them (the resulting binaries) available for cmake configuration run for target arch and build all those libraries and everything after the codegens run on target arch.

Apple did a really good job with their approach to two archs and universal binaries + cmake supported it, so on Apple you almost don't think about archs at all, you build all dependencies for both archs and join them in universal binary libraries (the same way on any arch you build, on Intel or Apple Silicon, you build them the same way) and then you build half of your targets (libraries and codegens) for both archs the same way on both hardware setups, and all target-arch-only libraries and tdesktop itself could be built either for both archs into universal binaries (if you need to deploy) or for a single arch only (if you need to debug and test, then you build only your host arch) -- and again, you do it the same way on any hardware, Intel or Apple Silicon. And cmake helps you all the way, giving you a way to configure wanted archs on per-target basis.

I think if Microsoft could've done something similar that would really improve the adoption of Windows on Arm for developers. I understand that's really hard thing to do so that it all works smoothly. But here we are. Another pain on Windows for that task is that some dependencies (current libvpx and ffmpeg) have no way to be built natively in development command prompt and rely on msys2 to be built for Windows, which adds complexity to building for different archs.

**Comment by CryptoManiac** [comment]
> I have Windows on Arm in Parallels Studio there. I just didn't manage to go further than first dependency out of 28 that I need to compile in order to build tdesktop on Arm 😢 

I've gave up on 11th or thirteen, don't remember exactly. Stumbled upon x86 inline assembly somewhere and decided to act like that exorcist priest from a movie. Some WebRTC library if I remember correctly.

**Comment by sungaila** [comment]
Thanks for the in-depth explanation, it makes it clearer which hurdles have to be overcome. Thankfully the x64 emulation is good enough imho.

**Comment by john-preston** [comment]
@CryptoManiac I got some problems right on msys2 part, if it could be overcome (building of ffmpeg and libvpx for arm native) then perhaps I could move further. The webrtc is built using my own cmake configuration from https://github.com/desktop-app/tg_owt, so I could patch it to build, I guess, because it builds fine for Arm on macOS.

**Comment by CryptoManiac** [comment]
> @CryptoManiac I got some problems right on msys2 part, if it could be overcome (building of ffmpeg and libvpx for arm native) then perhaps I could move further. The webrtc is built using my own cmake configuration from https://github.com/desktop-app/tg_owt, so I could patch it to build, I guess, because it builds fine for Arm on macOS.

I did all steps manually without using this script. Got too much trouble with it.

**Comment by NicolasOchoaMSFT** [comment]
> I have Windows on Arm in Parallels Studio there. I just didn't manage to go further than first dependency out of 28 that I need to compile in order to build tdesktop on Arm 😢
> 
> The cross-compiling also sucks for tdesktop because there are several code generator built and used (!) before tdesktop itself, so the resulting process is very complicated if you try to cross-compile. You need to build some libraries and codegens in host arch, then make them (the resulting binaries) available for cmake configuration run for target arch and build all those libraries and everything after the codegens run on target arch.
> 
> Apple did a really good job with their approach to two archs and universal binaries + cmake supported it, so on Apple you almost don't think about archs at all, you build all dependencies for both archs and join them in universal binary libraries (the same way on any arch you build, on Intel or Apple Silicon, you build them the same way) and then you build half of your targets (libraries and codegens) for both archs the same way on both hardware setups, and all target-arch-only libraries and tdesktop itself could be built either for both archs into universal binaries (if you need to deploy) or for a single arch only (if you need to debug and test, then you build only your host arch) -- and again, you do it the same way on any hardware, Intel or Apple Silicon. And cmake helps you all the way, giving you a way to configure wanted archs on per-target basis.
> 
> I think if Microsoft could've done something similar that would really improve the adoption of Windows on Arm for developers. I understand that's really hard thing to do so that it all works smoothly. But here we are. Another pain on Windows for that task is that some dependencies (current libvpx and ffmpeg) have no way to be built natively in development command prompt and rely on msys2 to be built for Windows, which adds complexity to building for different archs.

@john-preston  Hey John, thnak you so much for your interest and for providing your concerns for your development efforts. We are very interested in working together to make sure Telegram is supported on Windows on Arm devices. Could you please reach out to our team's email address, i will reach out directly to you so we can have a sync call with our dev engineers to go over the main approach to collaborate. Please send us an email to achelp@microsoft.com , once we receive the email i will reach out to you directly or not sure if you can reach out to me through  DM .  And we will be more than glad to send you a Windows Dev Kit directly to you for your development efforts.

**Comment by NicolasOchoaMSFT** [comment]
@john-preston  Hi John, hope you are well. Just following up on this thread, we are ready to help and collaborate together for Telegram to be supported on Windows on Arm devices. As advised, our service is completely free of charge and we have specialized developers that will work with you at any time to provide guidance and troubleshooting if needed. We have a Dev Kit ready for shipment as well to make sure your testing and development process is as efficient as possible.

**Comment by john-preston** [comment]
@NicolasOchoaMSFT I've successfully built it 😆 Although it was very hard with dependencies.. I'll drop an email anyway.

**Comment by NicolasOchoaMSFT** [comment]
> @NicolasOchoaMSFT I've successfully built it 😆 Although it was very hard with dependencies.. I'll drop an email anyway.

Hi John, hope you are well. Amazing news! we are very glad to hear these news. We have not received the email, i wil be pending on this so we can have a quick sync on how we can collaborate together as we are super excited to onboard Telegram for Windows on Arm devices. Thank you again, looking forward to talk further.

**Comment by john-preston** [comment]
Ok, everyone. So, I really went and invested significant amount of time in building all the dependencies (30 now) for ARM64 natively and built the TDesktop app itself and it was working fine (apart from the OpenGL parts fuck up due to Qt 6 which should be dealt with later), and even patched my release building scripts to support preparing / updating a Windows on ARM build, and decided to build 5.2.3 for a test, and...

![image](https://github.com/user-attachments/assets/0f7c561e-d607-4408-b336-188ac315a018)

Well, this is a stop for now. I have no idea, how to report that, because it is the final linking step and to reproduce that it requires building all the dependencies and the app itself, which will take a couple of hours on good hardware, I guess.

**Comment by CryptoManiac** [comment]
> Ok, everyone. So, I really went and invested significant amount of time in building all the dependencies (30 now) for ARM64 natively and built the TDesktop app itself and it was working fine (apart from the OpenGL parts fuck up due to Qt 6 which should be dealt with later), and even patched my release building scripts to support preparing / updating a Windows on ARM build, and decided to build 5.2.3 for a test, and...
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/0f7c561e-d607-4408-b336-188ac315a018)
> 
> 
> 
> Well, this is a stop for now. I have no idea, how to report that, because it is the final linking step and to reproduce that it requires building all the dependencies and the app itself, which will take a couple of hours on good hardware, I guess.

This is a compiler bug. Got this and similar errors many times in various random scenarios on all supported platforms. Didn't really notice any pattern. Sometimes the change of compiler settings and optimization level might help. Sometimes it doesn't. Most of the times it was dealt with by replacing the affected 3rd party code with the alternatives. I'm afraid there isn't much you can do to resolve this.

**Comment by john-preston** [comment]
Yes, I understand it's a compiler bug. But usually to report and fix them you need to provide a minimal reproducible example. Here it's hard to do, minimal example is whole project.

**Comment by ssg** [comment]
> Well, this is a stop for now. I have no idea, how to report that, 

Great that we have @NicolasOchoaMSFT here to help us. 

Would you like to put your working draft on some unofficial repo so we can check it out and try to fix it too?

**Comment by pbo-linaro** [comment]
You may give a try compiling with `clang-cl`, instead of msvc `cl` (included in windows-arm64 latest llvm release: [18.1.0](https://github.com/llvm/llvm-project/releases/download/llvmorg-18.1.0/LLVM-18.1.0-woa64.exe)).
It's *mostly* compatible (command line, ABI and dialects), so the switch should be easy.
Just add bin folder to your PATH and specify `-DCMAKE_CXX_COMPILER=clang-cl -DCMAKE_C_COMPILER=clang-cl` to cmake command line.

Note, if you cross compile from x64, you need to add a c/c++ flag to set target: `clang-cl -target aarch64-pc-windows-msvc`.

**Comment by NicolasOchoaMSFT** [comment]
@john-preston I have reached out to your email to confirm our engagement. We can report this issue you are facing to our engineers in a support ticket and we can proceed with the device shipment for further testing and debugging.

Looking forward to any follow up from your side. Thank you

**Comment by P40L0X** [comment]
> @john-preston I have reached out to your email to confirm our engagement. We can report this issue you are facing to our engineers in a support ticket and we can proceed with the device shipment for further testing and debugging.
> 
> Looking forward to any follow up from your side. Thank you

If you'll be able to compile it (and I really hope so), will be possible in the future to update/upgrade all current x64 builds officially installed through MS Store to a new native ARM64 version automatically?
Or people would need to uninstall current emulated one first?

Thanks,
-P

**Comment by john-preston** [comment]
@P40L0X Well, I could upload a new build to the store, I have no idea how the store will update the existing installations.

**Comment by P40L0X** [comment]
> @P40L0X Well, I could upload a new build to the store, I have no idea how the store will update the existing installations.

I hope we'll find out soon... 🙂

**Comment by P40L0X** [comment]
> @P40L0X Not really, this version (Windows build on Qt 6) requires huge amount of work regarding the media viewer, it is almost completely broken in current code :(

What about shifting to the WebApp (K) variant as a new base for the Windows Desktop app?
It seems more modern/fluid with better animations and upload speeds (but as just a web wrapper as of now, notifications and badges are a bit broken when using it at the moment)

**Comment by ilya-fedin** [comment]
@P40L0X you can download the desktop version of Web A from https://github.com/Ajaxy/telegram-tt/releases

**Comment by P40L0X** [comment]
> @P40L0X you can download the desktop version of Web A from https://github.com/Ajaxy/telegram-tt/releases

Thanks, but there's no native ARM64 version for Windows either...

**Comment by ilya-fedin** [comment]
> Thanks, but there's no native ARM64 version for Windows either...

I see Telegram-A-arm64.zip file there. Isn't this it?

**Comment by P40L0X** [comment]
> > Thanks, but there's no native ARM64 version for Windows either...
> 
> I see Telegram-A-arm64.zip file there. Isn't this it?

Nope, that's also for Mac OS X.

**Comment by ilya-fedin** [comment]
Ok. So, if you want arm64 build of Web A, you have to raise that on its repo. If you want desktop builds for Web K, you have to raise that on Web K repo (https://github.com/morethanwords/tweb).

**Comment by P40L0X** [comment]
> Ok. So, if you want arm64 build of Web A, you have to raise that on its repo. If you want desktop builds for Web K, you have to raise that on Web K repo (https://github.com/morethanwords/tweb).

Thanks, I opened a request for Telegram Web A.

**Comment by ilya-fedin** [comment]
Doesn't Unigram fit your needs? It seem to have ARM64 build...

**Comment by P40L0X** [comment]
> Doesn't Unigram fit your needs? It seem to have ARM64 build...

They used to have a native ARM64 build but they gave up and reverted to x64 again...

**Comment by ilya-fedin** [comment]
Oh. Given how it seem to be problematic for every Telegram client over there, it feels like the Windows ARM64 ecosystem is just not ready...

**Comment by sungaila** [comment]
> Oh. Given how it seem to be problematic for every Telegram client over there, it feels like the Windows ARM64 ecosystem is just not ready...

I the case of Unigram the devs are just unwilling to support a "~0.006% userbase".

This whole x64 -> ARM64 transition reminds me of x86 -> x64 back in the days of Vista: "Why should I port my app to x64? The x86 build runs just fine on your machine and almost no one has a x64 CPU and OS anyway."

**Comment by grandrew** [comment]
Good news!

Just installed Telegram with `sudo snap install telegram-desktop` 

Works flawlessly, super snappy, callling works as expected with an instant connection. The only two things I noticed so far are:

1. WSL consumes 3+GB memory (not sure if it's resident and/or how it compares to other windows components though so could be not an issue)
2. It can't access my Surface camera, but otherwise video calling also works as expected

![image](https://github.com/user-attachments/assets/cdac9e84-808d-4d68-80e9-ee94afbe7f4d)

![image](https://github.com/user-attachments/assets/1dacb8f1-254f-44cd-af3f-6c9c6c904e90)

**Comment by ilya-fedin** [comment]
> I the case of Unigram the devs are just unwilling to support a "~0.006% userbase".

From my reading of  https://github.com/UnigramDev/Unigram/issues/3010 they have problems with the same deps as tdesktop (various video codecs)  so the cause seem to be the same.

And, yes, the cause if this deps problem is likely the little userbase. If it would be big enough, it would accelerate development at all points: tdesktop, unigram, third party deps. As of now, there seem to be no way to justify the huge amount of work in all of: tdesktop (porting Windows-specific code to Qt 6, patching third party deps), Unigram (patchin third party deps), third party deps (fixing the bugs Unigram had to patch).

So it seems we have to just wait until ARM64 becomes really popular, at which point it will be easy for Unigram to switch. Not sure about tdesktop though, as porting to Qt 6 won't happen on its own unlike the deps being fixed.

**Comment by ilya-fedin** [comment]
So @john-preston finally did it after that public message 😅
The results aren't good though. Linux arm64 users are around 0.001% of all Linux users.

**Comment by ssg** [comment]
@ilya-fedin how much are Windows x86 users on Windows ARM64 though?

**Comment by ilya-fedin** [comment]
@ssg there's no such information

**Comment by ssg** [comment]
@ilya-fedin according to @john-preston's comment on Unigram stats above, that's ~0.006%. 0.007% in total is miniscule, but these numbers predate the Copilot PC launch two months ago. Apparently, [Copilot PCs made 20% of all PC sales in their launch week](https://www.digitaltrends.com/computing/battery-life-driving-copilot-sales/). 

So, expect these numbers to increase, but that's a bit of a chicken and egg problem too. I'm sure people can back away from ARM64 if "Telegram is so slow on it".

**Comment by ssg** [comment]
@CryptoManiac "Obviously manipulated", how do you figure that? AMD Ryzen AI isn't part of Copilot PC lineup and isn't part of Copilot PC launch statistics: https://www.microsoft.com/en-us/windows/copilot-plus-pcs

**Comment by CryptoManiac** [comment]
> @CryptoManiac "Obviously manipulated", how do you figure that? AMD Ryzen AI isn't part of Copilot PC lineup and isn't part of Copilot PC launch statistics: https://www.microsoft.com/en-us/windows/copilot-plus-pcs

First of all, I'm too used to seeing Microsoft lying about something and manipulating their "statistics" so it won't be too unexpected for them to do that once again.

I'm not sure whether you have noticed this so far, but it's August now. First ZenBooks came to market more than a month ago.

**Comment by marcpems** [comment]
> @P40L0X Well, I could upload a new build to the store, I have no idea how the store will update the existing installations.

The Store will update based on CPU architecture within version number, so if a newer version is available and there is an Arm64 and X64 package, the Store will deliver Arm64 to an Arm powered Windows machine. 
End result is Arm users will just get the update when its available.

**Comment by marcpems** [comment]
> m not sure whether you have noticed this so far, but it's August now. First Ryzen AI ZenBooks came to market more than two weeks ago. Then there will be an intel's lunar lake too. These ARM based Copilot PCs are practically doomed to be a minority after that. Their performance and battery life aren't that advanced to make someone choose them over x86-based Copilot PCs.

Today there are only Arm powered Copilot+ PCs available.

**Comment by ssg** [comment]
> First of all, I'm too used to seeing Microsoft lying about something and manipulating their "statistics" so it won't be too unexpected for them to do that once again.

Can you give me an example where Microsoft manipulated their statistics before?

> I'm not sure whether you have noticed this so far, but it's August now. First Ryzen AI ZenBooks came to market more than two weeks ago. Then there will be an intel's lunar lake too. These ARM based Copilot PCs are practically doomed to be a minority after that. Their performance and battery life aren't that advanced to make someone choose them over x86-based Copilot PCs.

I wish nothing more than having power efficient x86 PCs, but I gotta see it first. As you said, it can kill ARM64 lineup, or who knows, it can push the competition. What really matters is that there are way more ARM64 PCs in use now than the time of the quoted numbers.

**Comment by ilya-fedin** [comment]
Looks like I forgot to multiple the number by 100. It's 0.1% actually, fixed the initial message.

**Comment by ilya-fedin** [comment]
It's very sad in absolute numbers though. Will tell if get permission from @john-preston.

**Comment by P40L0X** [comment]
> > First of all, I'm too used to seeing Microsoft lying about something and manipulating their "statistics" so it won't be too unexpected for them to do that once again.
> 
> Can you give me an example where Microsoft manipulated their statistics before?
> 
> > I'm not sure whether you have noticed this so far, but it's August now. First Ryzen AI ZenBooks came to market more than two weeks ago. Then there will be an intel's lunar lake too. These ARM based Copilot PCs are practically doomed to be a minority after that. Their performance and battery life aren't that advanced to make someone choose them over x86-based Copilot PCs.
> 
> I wish nothing more than having power efficient x86 PCs, but I gotta see it first. As you said, it can kill ARM64 lineup, or who knows, it can push the competition. What really matters is that there are way more ARM64 PCs in use now than the time of the quoted numbers.

Yes, there are (and will be) noticeable more ARM64 users now that Copilot+ PC with Snapdragon X Elite/Plus were adopted by the vast majority of popular brands as Lenovo, Dell, Asus, HP etc. other than Microsoft (with their Surface lineup).

Snapdragon X Elite also demonstrated to be more powerful and efficient than latest Intel Ultra chips and on par, CPU performance wise, to latest AMD AI chips but much more efficient.

They're actually on par with Apple M3 Pro 12-Core performance/efficiency which is an absolute first for Windows.

So: ARM64 on Windows' adoption rate will just go up from now on.

**Comment by CryptoManiac** [comment]
> Can you give me an example where Microsoft manipulated their statistics before?

I know that it's quite old but the whole Get the Facts campaign had been enough to make North Korea's official TV look like an example of the unbiased journalism. Didn't ever take them seriously ever since then.

> I wish nothing more than having power efficient x86 PCs, but I gotta see it first. As you said, it can kill ARM64 lineup, or who knows, it can push the competition. What really matters is that there are way more ARM64 PCs in use now than the time of the quoted numbers.

Well you can use Microsoft's own Surface then. These devices aren't anything different from their ARM equivalents. Same performance, same battery life. They even look the same. I've spent a month with the Intel-based Surface Pro 9 to compare it with my own Surface Pro 9 ARM and I fail to see a point of the whole affair. Seriously, software compatibility issues are truly the only difference I can see. The dimensions are the same, the screen is the same, and the Intel version is something like 10 grams lighter which isn’t a difference you’d be able notice. 

I know that SP9 is quite old too but the new Qualcomm CPUs aren't game changer too. They've managed to catch up with the base version of Apple M3 and making it look like there are no pro and max versions. That's it. With the Lunar Lake you're going to have the same situation, two architectures without any noticeable user experience. Nothing will change here unless Microsoft will stop making hot air and begin to take their own platform seriously. Like investing money into ARM64 startups and probably buying a few of them to create their own department. You know, like actually doing something. Making an impression of progress and actually doing something aren't the same thing.

**Comment by P40L0X** [comment]
> > Can you give me an example where Microsoft manipulated their statistics before?
> 
> I know that it's quite old but the whole Get the Facts campaign had been enough to make North Korea's official TV look like an example of the unbiased journalism. Didn't ever take them seriously ever since then.
> 
> > I wish nothing more than having power efficient x86 PCs, but I gotta see it first. As you said, it can kill ARM64 lineup, or who knows, it can push the competition. What really matters is that there are way more ARM64 PCs in use now than the time of the quoted numbers.
> 
> Well you can use Microsoft's own Surface then. These devices aren't anything different from their ARM equivalents. Same performance, same battery life. They even look the same. I've spent a month with the Intel-based Surface Pro 9 to compare it with my own Surface Pro 9 ARM and I fail to see a point of the whole affair. Seriously, software compatibility issues are truly the only difference I can see. The dimensions are the same, the screen is the same, and the Intel version is something like 10 grams lighter which isn’t a difference you’d be able notice.
> 
> I know that SP9 is quite old too but the new Qualcomm CPUs aren't game changer too. They've managed to catch up with the base version of Apple M3 and making it look like there are no pro and max versions. That's it. With the Lunar Lake you're going to have the same situation, two architectures without any noticeable user experience. Nothing will change here unless Microsoft will stop making hot air and begin to take their own platform seriously. Like investing money into ARM64 startups and probably buying a few of them to create their own department. You know, like actually doing something. Making an impression of progress and actually doing something aren't the same thing.

My Lenovo Yoga Slim 7x 14" is actually better than Apple Macbook Pro 14" M3 and neck and neck with MBP 16" (with M3 Pro 12-Core) CPU wise while on par with AMD 780M iGPU performance wise (while consuming much less power and being silent) and all while costing half than MPB 32GB variant...

Windows 11 24H2 on ARM also much improved both ARM64 performance and x64 Emulation performance via Prism emulator and Windows itself already started leveraging the new NPUs for AI accelerated stuff.

Like I said, things really changed for the better and are finally getting serious with Windows on ARM this time, both hardware and software side.

So for third party apps it should be better adapting now rather than later or they could be superseeded by better alternatives altogether.

**Comment by CryptoManiac** [comment]
> Snapdragon X Elite also demonstrated to be more powerful and efficient than latest Intel Ultra chips and on par, CPU performance wise, to latest AMD AI chips but much more efficient.

The most efficient variants of X Plus and X Elite are those that are least clocked. The best efficiency is reached with 20-25 W TDP cap. Which means that you've got to choose either efficiency or performance and you can't have both. It's quite a dilemma. Then here comes a fact that they're not really that efficient  while significantly slower.

Cinebench R23 (Single-Core)

Snapdragon X Elite (X1E-84-100)
1804
Ryzen AI 9 HX 370 +13%
2041

Cinebench R23 (Multi-Core)

Snapdragon X Elite (X1E-84-100)
14837
Ryzen AI 9 HX 370 +62%
23972

Take these 62 percent multi core performance difference and amend them with a fact that Ryzen AI 9 HX 370 is merely 10 percent worse in terms of efficiency. And better to not even think about comparing the graphics. The Snapdragon GPU will be literally smashed into pieces.

> They're actually on par with Apple M3 Pro 12-Core performance/efficiency while costing MUCH less, which is an absolute first for Windows.

That's not good news either. First of all, Apple M3 is not different from the M2 both performance and efficiency wise, because it's essentially M2 on the N3b process. The very N3b process which is old enough to not be used for anything else and is getting phased out right now. Comparing whatever product with them is a manipulation and trying to make things look better than they are. You can forget about the M3 line entirely because it's a dead end. The M4 like is a new design which has nothing to do with M2/M3 line and it's them they've got to compete with.

Then there comes a fact that they're not that equal. They're on par when it comes to the multicore performance but roughly 15-20 percent slower in single core performance tests. Not mentioning that Snapdragon graphics performance is ridiculous. Like twice slower or something.

**Comment by ssg** [comment]
> I know that it's quite old but the whole Get the Facts campaign 

I'd say 20+ years of clean track record is a good sign, not a bad one.

> Well you can use Microsoft's own Surface then.

Surface ARM64 still has significantly longer battery life than its x86 equivalent.

That said, this is quite getting off-topic. Maybe we should move "Is ARM64 any good?" discussions to somewhere else and keep this thread to Telegram porting efforts.

**Comment by P40L0X** [comment]
> > Snapdragon X Elite also demonstrated to be more powerful and efficient than latest Intel Ultra chips and on par, CPU performance wise, to latest AMD AI chips but much more efficient.
> 
> The most efficient variants of X Plus and X Elite are those that are least clocked. The best efficiency is reached with 20-25 W TDP cap. Which means that you've got to choose either efficiency or performance and you can't have both. It's quite a dilemma. Then here comes a fact that they're not really that efficient while significantly slower.
> 
> Cinebench R23 (Single-Core)
> 
> Snapdragon X Elite (X1E-84-100) 1804 Ryzen AI 9 HX 370 +13% 2041
> 
> Cinebench R23 (Multi-Core)
> 
> Snapdragon X Elite (X1E-84-100) 14837 Ryzen AI 9 HX 370 +62% 23972
> 
> Take these 62 percent multi core performance difference and amend them with a fact that Ryzen AI 9 HX 370 is merely 10 percent worse in terms of efficiency. And better to not even think about comparing the graphics. The Snapdragon GPU will be literally smashed into pieces.
> 
> > They're actually on par with Apple M3 Pro 12-Core performance/efficiency while costing MUCH less, which is an absolute first for Windows.
> 
> That's not good news either. First of all, Apple M3 is not different from the M2 both performance and efficiency wise, because it's essentially M2 on the N3b process. The very N3b process which is old enough to not be used for anything else and is getting phased out right now. Comparing whatever product with them is a manipulation and trying to make things look better than they are. You can forget about the M3 line entirely because it's a dead end. The M4 line is a new design which has nothing to do with M2/M3 line and that's what they've truly got to compete with.
> 
> Then there comes a fact that they're not that equal. They're on par when it comes to the multicore performance but roughly 15-20 percent slower in single core performance tests. Not mentioning that Snapdragon graphics performance is ridiculous. Like twice slower or something.

Cinebench R23 is not native ARM64 so those numbers are meaningless (same as "graphics performance twice slower or something" you mentioned. That's because it's emulated.

You need to see Cinebench R24 numbers and other graphics numbers for native ARM64 benchmarks to actually measure performance/efficiency ratio.

Like I said we're in M3 Pro territory at half the price with those Snapdragons and that's a fact.

Now, I agree it's better to go back on topic: comment the need to have a local/native ARM64 version of Telegram Desktop which is actually one of the very few popular apps still not providing a native version at this point.

**Comment by pbo-linaro** [comment]
@john-preston Since you've been the one to do concrete work on this port, which bits are missing today? Have you been able to give a try to clang-cl to go around msvc bug?

**Comment by ilya-fedin** [comment]
> Have you been able to give a try to clang-cl to go around msvc bug?

I can't count how much times I proposed him to try clang-cl when msvc was bugging even on x86/x86_64. He seem to not to like changing the compiler on any platform. So I'm afraid we would know whether it helps only if someone else interested from the community would try.

**Comment by ilya-fedin** [comment]
I guess his strategy is waiting when the ICE would be reported by someone else to MSVC and fixed (apparently trying again e.g. after months/a year or something like that).

**Comment by john-preston** [comment]
@ilya-fedin Nope, I've worked with a Microsoft employee to provide him a repro for the linking crash, we found it was fixed in the Preview and found another one there and I've produced a repro for that another one and he reproduced and told me he fixed it privately and was working to push the fix in some upcoming release of msvc and will be in touch once he gets some news.

**Comment by pbo-linaro** [comment]
Thanks for your update @john-preston. Do you have a public branch where your changes might be available so I can try to reproduce on my side too?

**Comment by ilya-fedin** [comment]
Were there any changes? If I'm not mistaken, he was building the current code, just running prepare.py lots of times until it succeeds. Although you likely need to build with Qt 6 which is undocumented but only a question of passing additional arguments to prepare.py and configure.py.

**Comment by ilya-fedin** [comment]
If you have some concrete problem when building, error messages would likely lead to more concrete suggestions.

**Comment by john-preston** [comment]
@pbo-linaro Well, I have pushed the changes required for dependencies building already. Now the Debug version can be built and a lot of problems need to be dealt with.

**Comment by pbo-linaro** [comment]
Pushed on `dev` branch, or in a fork of yours?

**Comment by rorychatt** [comment]
Please make this thing as ARM is the future standard for all apps!

**Comment by lexcyn** [comment]
> @pbo-linaro Well, I have pushed the changes required for dependencies building already. Now the Debug version can be built and a lot of problems need to be dealt with.

I was able to get as far as compiling the libraries, but running the config for ARM tells me that there is no QT version installed, right after selecting the correct QT version. Trying to determine why:

`Choosing Qt 5.
Choosing Qt 5.15.13
Cleared previous.
Warn about uninitialized values.
-- Building for: Visual Studio 17 2022
Version: 5.4.1
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.27686.
-- The C compiler identification is MSVC 19.41.34120.0
-- The CXX compiler identification is MSVC 19.41.34120.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostarm64/arm64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostarm64/arm64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at cmake/external/qt/package.cmake:56 (message):
  Neither Qt6 nor Qt5 is found
Call Stack (most recent call first):
  CMakeLists.txt:60 (include)


-- Configuring incomplete, errors occurred!
FAILED`

**Comment by john-preston** [comment]
@lexcyn Try running "./build/prepare/win.bat qt6" to build Qt 6 and then run "./configure.bat qt6" to configure for Qt 6.

**Comment by lexcyn** [comment]
> @lexcyn Try running "./build/prepare/win.bat qt6" to build Qt 6 and then run "./configure.bat qt6" to configure for Qt 6.

Running the win.bat with qt6 skipped (since qt6 is already installed and configured) and running configure pointing to qt6 did work as now I see:

```
Choosing Qt 6.
Choosing Qt 6.7.2
Cleared previous.
```
However I get the same error right after:

```
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at cmake/external/qt/package.cmake:56 (message):
  Neither Qt6 nor Qt5 is found
Call Stack (most recent call first):
  CMakeLists.txt:60 (include)
-- Configuring incomplete, errors occurred!
FAILED
```

**Comment by ilya-fedin** [comment]
Are you sure prepare.py is successful? Assuming everything is right, you should have `Libraries` folder near `tdesktop` folder and `Qt-6.7.2` folder in the `Libraries` folder.

**Comment by lexcyn** [comment]
> Are you sure prepare.py is successful? Assuming everything is right, you should have `Libraries` folder near `tdesktop` folder and `Qt-6.7.2` folder in the `Libraries` folder.

Yep it ran without error and I do see the qt-6.7.2 folder. The only difference to the regular build instructions is I am using the ARM64 native toolchain compiler and instead of using the 'x64' argument on configure.bat I am using 'arm'

**Comment by ilya-fedin** [comment]
> qt-6.7.2

With that exact casing? Not Qt-6.7.2?

**Comment by lexcyn** [comment]
> > qt-6.7.2
> 
> With that exact casing? Not Qt-6.7.2?

No it's Qt-6.7.2

<img width="155" alt="image" src="https://github.com/user-attachments/assets/18bd9bb1-312b-4fba-96e8-50ca8ea00d13">

**Comment by ilya-fedin** [comment]
Does it have `lib\cmake\Qt6Config.cmake` file?

**Comment by ilya-fedin** [comment]
> Assuming everything is right, you should have `Libraries` folder near `tdesktop` folder

And please confirm that part, that it is near `tdesktop` folder.

**Comment by lexcyn** [comment]
> Does it have `lib\cmake\Qt6Config.cmake` file?

Yes but it's under lib\cmake\Qt6\Qt6Config.cmake

> > Assuming everything is right, you should have `Libraries` folder near `tdesktop` folder
> 
> And please confirm that part, that it is near `tdesktop` folder.

Yes I have my build directory set up like this:

<img width="134" alt="image" src="https://github.com/user-attachments/assets/4c532157-3fc4-4a28-a0f4-6eee9cd3fa82">

**Comment by ilya-fedin** [comment]
Have you tried to delete `out` folder?

**Comment by ilya-fedin** [comment]
Can you upload your `out/CMakeCache.txt`?

**Comment by lexcyn** [comment]
> Can you upload your `out/CMakeCache.txt`?

Sure (I removed API info):

[CMakeCache.txt](https://github.com/user-attachments/files/16801407/CMakeCache.txt)

**Comment by ilya-fedin** [comment]
I assume removing QUIET from those lines should reveal more information: https://github.com/desktop-app/cmake_helpers/blob/master/external/qt/package.cmake#L51-L54

tdesktop build system has a strongly verified list of Qt deps while Qt uses autodetection. If Qt has found some library on your computer, it could build with it while tdesktop not supporting that so the configuration will fail unless that library is removed and Qt is rebuilt.

**Comment by lexcyn** [comment]
> I assume removing QUIET from those lines should reveal more information: https://github.com/desktop-app/cmake_helpers/blob/master/external/qt/package.cmake#L51-L54
> 
> tdesktop build system has a strongly verified list of Qt deps while Qt uses autodetection. If Qt has found some library on your computer, it could build with it while tdesktop not supporting that so the configuration will fail unless that library is removed and Qt is rebuilt.

Thanks - removed QUIET and now see:

```
--Could NOT find QT (missing: QT_DIR)
--Could NOT find QT (missing: QT_DIR)
```

I did re-run the pre-req build again but it doesn't give any errors and detects all libraries properly - I'm wondering if there is something with the path for QT not being set properly for ARM?

**Comment by lexcyn** [comment]
OK I figured out the QT issue - for whatever reason it was not added to PATH. Updated to add QTDIR=my path to Qt and it now gets further, however it looks like cmake is looking for the wrong arch for all libraries (external_qt is looking for win64 directory but the actual dir is just `TGRAM/Libraries/win64/Qt-6.7.2/include`:

```
CMake Error in cmake/external/kimageformats/CMakeLists.txt:
  Imported target "external_qt" includes non-existent path

    "C:/Users/devin/Downloads/TGRAM/Libraries/win64/Qt-6.7.2/include"
```

**Comment by lexcyn** [comment]
> @lexcyn looks like all the issue is that the check on this line fails for you: https://github.com/desktop-app/cmake_helpers/blob/master/variables.cmake#L78

You're right - that should not be `arm` but rather `arm64` - according to cmake it uses the `PROCESSOR_ARCHITECTURE` system variable which on my system at least is set to `ARM64`

**Comment by ilya-fedin** [comment]
> You're right - that should not be `arm` but rather `arm64`

Can you confirm it helps? I guess it could still not work if it's case sensitive.

**Comment by ilya-fedin** [comment]
Although maybe it's meant to be `arm.*` given that it uses `MATCHES` rather than `STREQUAL`.

**Comment by lexcyn** [comment]
> Although maybe it's meant to be `arm.*` given that it uses `MATCHES` rather than `STREQUAL`.

Let me give that a try in a few - just rebuilding the libraries as I had some errors with the final VS compile (_possible I messed something up when I was trying to resolve the previous errors_)

**Comment by lexcyn** [comment]
> Although maybe it's meant to be `arm.*` given that it uses `MATCHES` rather than `STREQUAL`.

So it looks like it is case sensitive so lower case `arm*` did not work but `ARM*` did work

**Comment by lexcyn** [comment]
@NicolasOchoaMSFT @pbo-linaro success! Here's my guide on how to get it compiled. By the way, the speed/performance difference is quite staggering in the GUI alone (in a good way):

Follow the regular build guide except the following changes:

- Run the ARM64 Native Tools Command Prompt
- Use Python 3.11
- Add `QTDIR` into your environment variables (PATH) with the following location: `builddir\TGRAM\Libraries\Qt-6.7.2`
- Edit the tdesktop\cmake\variables.cmake file and adjust the following:
-         `if (CMAKE_SYSTEM_PROCESSOR MATCHES "arm")` change to `if (CMAKE_SYSTEM_PROCESSOR MATCHES "ARM*")`
- Change the arch in the build step to: `configure.bat arm qt6 -D TDESKTOP_API_ID=YOUR_API_ID -D TDESKTOP_API_HASH=YOUR_API_HASH`
- Note that I had to specify `qt6` in order to get the build to work but you may not need to

**Comment by ilya-fedin** [comment]
> * Add `QTDIR` into your environment variables (PATH) with the following location: `builddir\TGRAM\Libraries\Qt-6.7.2`

This shouldn't be needed

**Comment by lexcyn** [comment]
> > * Add `QTDIR` into your environment variables (PATH) with the following location: `builddir\TGRAM\Libraries\Qt-6.7.2`
> 
> This shouldn't be needed

It shouldn't but for whatever reason my build would just not detect QT unless I did this

**Comment by ilya-fedin** [comment]
Try now again.

**Comment by lexcyn** [comment]
> Try now again. I'm pretty sure that's a mis-observation and setting it will lead to problems in future (e.g. when the Qt version changes).

I'll give it a go and yes this is just what I did to get it working and should probably not be kept this way. I will edit my post if it works after removing it.

**Comment by ssg** [comment]
Great news! What's the next step to make it official?

**Comment by ilya-fedin** [comment]
@lexcyn can you try `CMAKE_SYSTEM_PROCESSOR MATCHES "ARM"`? I had a private talk with @john-preston and he wants it to be like that but it seems he has no time to try whether it works, so if you would confirm it works, I'd make a PR with the change.

**Comment by ilya-fedin** [comment]
> Great news! What's the next step to make it official?

There seem to be various major bugs in the build (e.g. broken media viewer) and it seems @john-preston has little time to work on that so this could take quite a lot time apparently.

**Comment by ilya-fedin** [comment]
I even have a fix for the media viewer bug but @john-preston has no time to review it.

**Comment by lexcyn** [comment]
> > Great news! What's the next step to make it official?
> 
> There seem to be various major bugs in the build (e.g. broken media viewer) and it seems @john-preston has little time to work on that so this could take quite a lot time apparently. They seem to be mostly related to Qt 6 port rather than ARM build though but ARM build depends on Qt 6 port.

Strange, I am still using 5.4.4 (which was on QT 6.7.2) and nothing is broken, it all works very well even the media player. But yes on the newer builds with QT 6+, the GUI feels laggy and it does crash on UI scaling (could be other crashes too, I will see if I can find any)

Also I did try `CMAKE_SYSTEM_PROCESSOR MATCHES "ARM"` and it did work!

**Comment by ilya-fedin** [comment]
> Strange, I am still using 5.4.4 (which was on QT 6.7.2) and nothing is broken

is it transparent like in the qt 5 version? don't you have bugs resizing it if disabling opengl? (although the latter seem to be reproducible on qt5 too)

> Also I did try `CMAKE_SYSTEM_PROCESSOR MATCHES "ARM"` and it did work!

Nice!

**Comment by ilya-fedin** [comment]
There's also a question what to do about OpenGL. OpenGL drivers are of very bad quality at mass level and it seems @john-preston doesn't want to release a stable version using it. Given that Qt 6 no more supports ANGLE (which translates OpenGL into D3D11 or D3D9), alternative solution is needed. The most obvious solution is to rewrite OpenGL code to the new Qt's RHI but that would require LOTS of time. There are Krita's WIP patches for Qt 6 to bring ANGLE back but @john-preston ignore when I sent him the link so I have no idea what he thinks about that.

**Comment by ilya-fedin** [comment]
@lexcyn I see your https://github.com/desktop-app/cmake_helpers/pull/360, will you update it?

**Comment by lexcyn** [comment]
> @lexcyn I see your [desktop-app/cmake_helpers#360](https://github.com/desktop-app/cmake_helpers/pull/360), will you update it?

Yes I will do that now!

**Comment by lexcyn** [comment]
> > Strange, I am still using 5.4.4 (which was on QT 6.7.2) and nothing is broken
> 
> is it transparent like in the qt 5 version? don't you have bugs resizing it if disabling opengl? (although the latter seem to be reproducible on qt5 too)
> 
> > Also I did try `CMAKE_SYSTEM_PROCESSOR MATCHES "ARM"` and it did work!
> 
> Nice!

Yes the background is transparent and resizing it does work in OpenGL mode and with OpenGL disabled.

**Comment by ilya-fedin** [comment]
> Yes the background is transparent and resizing it does work in OpenGL mode and with OpenGL disabled.

Ah, maybe this happens only with x86/x64... But anyway this has to be resolved for Windows Qt 6 port to happen.

**Comment by sungaila** [comment]
> Yes the background is transparent and resizing it does work in OpenGL mode and with OpenGL disabled.

I am wondering on which ARM hardware you have tested OpenGL. I got the older Snapdragon 8cx here and need the [OpenCL™, OpenGL®, and Vulkan® Compatibility Pack](https://apps.microsoft.com/detail/9nqpsl29bfff?hl=en-us) for that.

No idea about the newer Snapdragons (Copilot+PC) or Macs on virtualization though.

**Comment by ilya-fedin** [comment]
The effective state of the OpenGL could be seen in the log.txt. Or by trying to zoom the picture in the viewer (only OpenGL-based code has the animation).

**Comment by lexcyn** [comment]
> > Yes the background is transparent and resizing it does work in OpenGL mode and with OpenGL disabled.
> 
> Ah, maybe this happens only with x86/x64... But anyway this has to be resolved for Windows Qt 6 port to happen.

QT bug is being worked on now, I am going to test the code change listed here: https://codereview.qt-project.org/c/qt/qtbase/+/589214

**Comment by lexcyn** [comment]
Strange - unsure if this is an issue with QT or now with the app. Adjusting the UI scale "up/increase" gives a the v_top break at line 73 of qrunnable.h and adjusting it "down/smaller" gives a v_top break at line 4160 of qpaintengine_raster.cpp.

**Comment by ssg** [comment]
@lexcyn That QT fix seems to have been merged.

**Comment by lexcyn** [comment]
> @lexcyn That QT fix seems to have been merged.

Yes which is great, although there is still a bug with the v_top and v_bot variables - there's a warning during QT build and this causes MSVC breakpoint crashes when compiled. I've submitted another bug for QT to fix: https://bugreports.qt.io/browse/QTBUG-128850

**Comment by ilya-fedin** [comment]
But those are the same variables the previously linked fix changes. I.e. you reported the same thing second time.

**Comment by lexcyn** [comment]
> But those are the same variables the previously linked fix changes. I.e. you reported the same thing second time.



> But those are the same variables the previously linked fix changes. I.e. you reported the same thing second time.

Yes but it seems like it was not fully fixed. I had recreated my build environment and recompiled from scratch and still getting errors with those variables. At least now however I am getting warnings during the QT initial build and then breakpoints during GUI interactions after full Telegram compile (GUI scaling) with that same variable. I'm just basing that bug report on the results I'm seeing. I am not an expert in this by far so if I am way off base let me know :)

**Comment by ilya-fedin** [comment]
> I had recreated my build environment and recompiled from scratch and still getting errors with those variables.

Because you still build from the same Qt version where the bug is not fixed?

**Comment by ilya-fedin** [comment]
You would need something like that to get it applied
```
diff --git a/Telegram/build/prepare/prepare.py b/Telegram/build/prepare/prepare.py
index b05f65cded..c052ea2a7b 100644
--- a/Telegram/build/prepare/prepare.py
+++ b/Telegram/build/prepare/prepare.py
@@ -1667,6 +1667,7 @@ mac:
     ninja
     ninja install
 win:
+    git cherry-pick 998bffecc25dd735175d296303d18ed0df1f9651
     for /r %%i in (..\\..\\patches\\qtbase_%QT%\\*) do git apply %%i -v
     cd ..
 
```
But that's a pedantic warning rather than a real bug so you could just as well build tdesktop in release mode. It won't do those annoying runtime warnings and should work a lot faster.

**Comment by lexcyn** [comment]
> You would need something like that to get it applied
> 
> ```
> diff --git a/Telegram/build/prepare/prepare.py b/Telegram/build/prepare/prepare.py
> index b05f65cded..c052ea2a7b 100644
> --- a/Telegram/build/prepare/prepare.py
> +++ b/Telegram/build/prepare/prepare.py
> @@ -1667,6 +1667,7 @@ mac:
>      ninja
>      ninja install
>  win:
> +    git cherry-pick 998bffecc25dd735175d296303d18ed0df1f9651
>      for /r %%i in (..\\..\\patches\\qtbase_%QT%\\*) do git apply %%i -v
>      cd ..
>  
> ```
> 
> But that's a pedantic warning rather than a real bug so you could just as well build tdesktop in release mode. It won't do those annoying runtime warnings and should work a lot faster.

Dang you are right, I'm dumb - I will try and see if it works, I know the last time I tried a release build it failed to compile.

**Comment by sungaila** [comment]
> Looks like it's an issue with the compiler and Microsoft is working on a fix. https://developercommunity.visualstudio.com/t/Encountered-ICE-While-Building-Telegram-/10718860

[This could be one of the problems that John has already reported.](https://github.com/telegramdesktop/tdesktop/issues/25266#issuecomment-2292199532)

**Comment by lexcyn** [comment]
> > Looks like it's an issue with the compiler and Microsoft is working on a fix. https://developercommunity.visualstudio.com/t/Encountered-ICE-While-Building-Telegram-/10718860
> 
> [This could be one of the problems that John has already reported.](https://github.com/telegramdesktop/tdesktop/issues/25266#issuecomment-2292199532)

Great, hopefully it gets fixed soon!

**Comment by ilya-fedin** [comment]
Can anyone confirm that the right architecture is reported under emulation in system version on 5.5.7?

**Comment by ilya-fedin** [comment]
no one?

**Comment by lexcyn** [comment]
> no one?

I mean yes, the x64 version does run under emulation, just slower. The last build I was able to compile arm64 native was 5.4.4
<img width="225" alt="image" src="https://github.com/user-attachments/assets/fce117fe-6b6f-4ef2-b777-787088518621">

**Comment by ilya-fedin** [comment]
Huh? I meant what is in the system version here on the x64 build (not arm64) run on arm64 hardware
![](https://github.com/user-attachments/assets/e0165376-4ca0-4c6c-b736-e8569fd0686d)

**Comment by lexcyn** [comment]
> Huh? I meant what is in the system version here on the x64 build (not arm64) run on arm64 hardware ![](https://private-user-images.githubusercontent.com/17829319/373912645-e0165376-4ca0-4c6c-b736-e8569fd0686d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgxNzIwMzgsIm5iZiI6MTcyODE3MTczOCwicGF0aCI6Ii8xNzgyOTMxOS8zNzM5MTI2NDUtZTAxNjUzNzYtNGNhMC00YzZjLWI3MzYtZTg1NjlmZDA2ODZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA1VDIzNDIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiM2Q0OThhMmE3YmRiOTk2MzUzNzEyOGM5MWRjYmRmYTBhYTkyODQzNDEzNTIxOGU5ZTUwOWE4MjU3NWNjMDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pelocVbpIkpzHDY5qITkGw4H5mcuvNYXVF1fW-_NStU)

Sorry here you go:

<img width="263" alt="image" src="https://github.com/user-attachments/assets/bc0f2b48-18b7-47e8-af0b-141548970f66">

**Comment by ilya-fedin** [comment]
Wow, nice, so we can measure the amount of arm64 users now

**Comment by ilya-fedin** [comment]
Your arm64 build has arm64 in application string, right?

**Comment by lexcyn** [comment]
> Your arm64 build has arm64 in application string, right?

Yep it does:

<img width="197" alt="image" src="https://github.com/user-attachments/assets/da80159b-1512-441c-b751-3fcda47a2f85">

**Comment by ilya-fedin** [comment]
By the way, 5.6 now includes various Qt 6 network plugins on Windows, Qt 6.8.0-rc1 and the media viewer fixes. So the only remaining blockers should be the compiler bug and determining what to do with OpenGL.

**Comment by lexcyn** [comment]
> By the way, 5.6 now includes various Qt 6 network plugins on Windows, Qt 6.8.0-rc1 and the media viewer fixes. So the only remaining blockers should be the compiler bug and determining what to do with OpenGL.

That's great to hear. I have a case open with Microsoft regarding the compile bug as well, they are claiming a fix is coming soon.

**Comment by ArsKozlov** [comment]
Just received my Surface Laptop and am curious about aarch64 support of Telegram. I see some people manage to compile some versions. Could somebody summarize the current state of the issues now? Thanks

**Comment by lexcyn** [comment]
> Just received my Surface Laptop and am curious about aarch64 support of Telegram. I see some people manage to compile some versions. Could somebody summarize the current state of the issues now? Thanks

There's an issue with the compiler right now that Microsoft is trying to fix so while you can compile the dependencies etc the final compile won't work if you use the release build. If you try a debug build there are some odd QT debug errors as well but I haven't attempted this in a while. Last build I was able to successfully compile and use was 5.4.4 with QT 5.8 (which is what I am still using now)

**Comment by ArsKozlov** [comment]
> Last build I was able to successfully compile and use was 5.4.4 with QT 5.8 (which is what I am still using now)

I just started to compile dependencies. Could you share somehow your working build? Many thanks

**Comment by lexcyn** [comment]
> > Last build I was able to successfully compile and use was 5.4.4 with QT 5.8 (which is what I am still using now)
> 
> I just started to compile dependencies. Could you share somehow your working build? Many thanks

It's a debug build using my personal API key so I can't share it sorry!

**Comment by ArsKozlov** [comment]
Maybe you can help me with preparing sources?
I have some errors with ninja and armasm64 argument
![image](https://github.com/user-attachments/assets/2b5f620d-47ed-418a-a169-809d1229a115)

**Comment by lexcyn** [comment]
> Maybe you can help me with preparing sources? I have some errors with ninja and armasm64 argument ![image](https://private-user-images.githubusercontent.com/58520784/375895983-2b5f620d-47ed-418a-a169-809d1229a115.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2ODYxNjgsIm5iZiI6MTcyODY4NTg2OCwicGF0aCI6Ii81ODUyMDc4NC8zNzU4OTU5ODMtMmI1ZjYyMGQtNDdlZC00MThhLWExNjktODA5ZDEyMjlhMTE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDIyMzEwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxNTUwMzJiMzI3MTUzYzcyNThkYWM1ZDhlYjYwY2M1ZTVkMTNlZjkzN2E2ODc1M2MzZGNjYTkyMGUxODY5NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.23ObVHlr-f4ivEWhzvVjYu1jJFsEc6-Fwva9dFeQrvg)

You need to run in ARM64 native toolchain command prompt

**Comment by lexcyn** [comment]
Good news all! The debug errors I was getting are gone and I was able to compile the latest version in debug mode. Sadly the same compile error is still present trying to compile release. @ilya-fedin

**Comment by ArsKozlov** [comment]
> You need to run in ARM64 native toolchain command prompt

Of course, I used ARM64 Command Prompt. Probably, because I did it in 2022 Preview?
Maybe something wrong is in my environment. Trying now to reinstall to stable 2022

![image](https://github.com/user-attachments/assets/16ec2b13-895a-4b7e-9ca9-0ed8d5e6783e)

**Comment by ssg** [comment]
Visual Studio Preview version with the fix just got released: https://learn.microsoft.com/en-us/visualstudio/releases/2022/release-notes-preview?tabs=preview4#17.12.0-pre.4.0

**Comment by sungaila** [comment]
> Visual Studio Preview version with the fix just got released: https://learn.microsoft.com/en-us/visualstudio/releases/2022/release-notes-preview?tabs=preview4#17.12.0-pre.4.0

>#### Top bug fixes
>- [Encountered ICE While Building Telegram during LTCG for ARM64](https://developercommunity.visualstudio.com/t/Encountered-ICE-While-Building-Telegram-/10718860)

Awesome, one step closer! :-)

**Comment by john-preston** [comment]
Tried to build release version, got "heap exceeded" on 32 GB RAM 🤣 

The media viewer still is bugged 😞

**Comment by lexcyn** [comment]
> Tried to build release version, got "heap exceeded" on 32 GB RAM 🤣
> 
> The media viewer still is bugged 😞

I was able to build the release and it's working WAY better than debug (no more UI lag). Media viewer still works for me. What is bugged about it?

**Comment by ilya-fedin** [comment]
He said to me that it has glitches. Not very informative but I bet Parallels drivers are the cause (as he runs Windows ARM VM on macOS ARM machine to my knowledge)

**Comment by john-preston** [comment]
Ok, let's try this one out.

https://td.telegram.org/tarm64/tsetup-arm64.5.7.2.exe
https://td.telegram.org/tarm64/tportable-arm64.5.7.2.zip
https://github.com/telegramdesktop/tdesktop/releases/download/v5.7.2/tsetup-arm64.5.7.2.exe
https://github.com/telegramdesktop/tdesktop/releases/download/v5.7.2/tportable-arm64.5.7.2.zip

**Comment by sungaila** [comment]
> Ok, let's try this one out.
> 
> https://td.telegram.org/tarm64/tsetup-arm64.5.7.2.exe https://td.telegram.org/tarm64/tportable-arm64.5.7.2.zip https://github.com/telegramdesktop/tdesktop/releases/download/v5.7.2/tsetup-arm64.5.7.2.exe https://github.com/telegramdesktop/tdesktop/releases/download/v5.7.2/tportable-arm64.5.7.2.zip

Awesome, I works! 👍 I just clicked around in some chats and opened pictures/videos for testing.

![image](https://github.com/user-attachments/assets/dc8f51ed-f5f7-4d12-a77c-3b889507db3e)
![image](https://github.com/user-attachments/assets/0d2071ca-3e71-4623-8a04-a1333e15c3e0)
![image](https://github.com/user-attachments/assets/54159d27-ae04-4c8e-96b3-99186c704786)

Many thanks to all those who have worked to make this possible! 😄

**Comment by john-preston** [comment]
@sungaila Autoupdating wasn't tested yet, so this needs to be kept in mind.

**Comment by lexcyn** [comment]
> Ok, let's try this one out.
> 
> https://td.telegram.org/tarm64/tsetup-arm64.5.7.2.exe https://td.telegram.org/tarm64/tportable-arm64.5.7.2.zip https://github.com/telegramdesktop/tdesktop/releases/download/v5.7.2/tsetup-arm64.5.7.2.exe https://github.com/telegramdesktop/tdesktop/releases/download/v5.7.2/tportable-arm64.5.7.2.zip

Tried the setup and it's working fine here! Will this be added to the runners to auto build so we can test the auto update?

**Comment by ilya-fedin** [comment]
> Will this be added to the runners to auto build

No, github has no arm64 windows runner

> so we can test the auto update?

Runners aren't related to auto update, though. You just have to wait for a new version to test.

**Comment by lexcyn** [comment]
> > Will this be added to the runners to auto build
> 
> No, github has no arm64 windows runner
> 
> > so we can test the auto update?
> 
> Runners aren't related to auto update, though. You just have to wait for a new version to test.

I did see they are in pre-release but you're right, they haven't been released in production yet (https://github.com/actions/runner/releases)

I will definitely test on a couple of my ARM laptop devices when a new version comes out

**Comment by ilya-fedin** [comment]
Note that gh runners aren't related to auto updates nor tdesktop releases in general.

**Comment by ssg** [comment]
There is an official ARM64 release now! https://github.com/telegramdesktop/tdesktop/releases/tag/v5.7.2

Thank you everyone who worked on this!

**Comment by Vishnu-Qualcomm** [comment]
> > Ok, let's try this one out.
> 
> Awesome, I works! 👍 I just clicked around in some chats and opened pictures/videos for testing.
> 
> ![image](https://private-user-images.githubusercontent.com/12021771/383565855-dc8f51ed-f5f7-4d12-a77c-3b889507db3e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0MTI5NzQsIm5iZiI6MTczMTQxMjY3NCwicGF0aCI6Ii8xMjAyMTc3MS8zODM1NjU4NTUtZGM4ZjUxZWQtZjVmNy00ZDEyLWE3N2MtM2I4ODk1MDdkYjNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDExNTc1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzMGYzZTljYTE4NTNkMDY4MTFiYmFmYzQ4Yjk1MGIzODk2ZTJiYjcwOWViYmJiNzVkOTQwMjBhZDJjZmRlNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YlAAecYI9VJgRbAkxzvKXvZrzBbODO9H_aWh3rMxsHk) ![image](https://private-user-images.githubusercontent.com/12021771/383567835-0d2071ca-3e71-4623-8a04-a1333e15c3e0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0MTI5NzQsIm5iZiI6MTczMTQxMjY3NCwicGF0aCI6Ii8xMjAyMTc3MS8zODM1Njc4MzUtMGQyMDcxY2EtM2U3MS00NjIzLThhMDQtYTEzMzNlMTVjM2UwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDExNTc1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2YTAzNzI4MDMzYmQ3N2E3YmYwNjljNTk3NjRhYmRlYWQxMWYyYzM1NTdmMTcyMTM1NjE5NzhiZmI5MzgxMGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ku4Z5_19lWS1gGr_lBiSu4k-TsYU_Dv-87J0sd09uyY) ![image](https://private-user-images.githubusercontent.com/12021771/383566656-54159d27-ae04-4c8e-96b3-99186c704786.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0MTI5NzQsIm5iZiI6MTczMTQxMjY3NCwicGF0aCI6Ii8xMjAyMTc3MS8zODM1NjY2NTYtNTQxNTlkMjctYWUwNC00YzhlLTk2YjMtOTkxODZjNzA0Nzg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDExNTc1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmZTFiNDlkNDhlOTMyOTcwNzFjMjBhZDQzMDJiNmUwNTJkZjI5MDQ3MzQ4ZWNkZjY3ZWYwNmFkMmU3MjNhYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.czCc-jZpMUrvpckSx_opYlQPxynCFBD1GEiRJvu9i88)
> 
> Many thanks to all those who have worked to make this possible! 😄
> 
> Now I'm waiting for the update to appear in the Microsoft Store. So far tdesktop has been published as an appx there. Hopefully there will be no problems uploading both x64 and arm64 builds.

We would also like to know the official release ETA on the Microsoft Store. Please let us know if there are any plans for a release in the near future. We would appreciate any leads or information about it.

**Comment by lexcyn** [comment]
> @sungaila Autoupdating wasn't tested yet, so this needs to be kept in mind.

Tested auto updates today to 5.7.3 beta and it worked without error

**Comment by GeldHades27355** [comment]
Hey Team, first of all: Thanks for finally porting this to native Arm64. 👍 Thanks also for Qualcomm for supporting!

There is a BUG: As of today, I can't use "save as". The Windows file picker Window briefly opens but immediately closes again.

**Comment by sungaila** [comment]
@Vishnu-Qualcomm Telegram Desktop 5.8.1 is available in the Microsoft Store now. It even upgrades from the x64 build to the arm64 build on its own. Neat!

**Comment by Vishnu-Qualcomm** [comment]
> @Vishnu-Qualcomm Telegram Desktop 5.8.1 is available in the Microsoft Store now. It even upgrades from the x64 build to the arm64 build on its own. Neat!

Thanks @sungaila for the info, thanks to all the contributors and maintainers!

**Comment by john-preston** [comment]
@Vishnu-Qualcomm File dialogs indeed crash, I've fixed that in code and it'll be soon in the next version.

**Comment by Aokromes** [comment]
plz, for future issues open one issue for each bug, since this issue is already implemented.

**Comment by john-preston** [comment]
@Farce002 Hi, thanks. Can we skip the call? I'm not good with calls. But the build is in Windows Store now, so well done everyone, I guess.

**Comment by Farce002** [comment]
@john-preston sure that is not a problem at all, just one last thing, about Arm64 Support Statement, is that also available for general knowledge yet?

**Comment by Farce002** [comment]
@john-preston Thank you for working with the App Assure service for your application remediation needs. We have concluded efforts on this case. As per our conversation via Github the ARM64 build for Telegram Desktop is already available in MS Store, no closure call was needed.
< Hi, thanks. Can we skip the call? I'm not good with calls. But the build is in Windows Store now, so well done everyone, I guess.>
 
Upon your confirmation, I will close your App Assure request by Dec. 06. Please know that I will be readily available at any point to assist you with any future app compatibility issues that may arise. Feel free to reach out to me directly to re-engage. If you would like a refresher on what our service can provide, you can email us at [achelp@microsoft.com](mailto:achelp@microsoft.com).
​​​​​​​After closure, you will get an automated email from Fastrack Support titled “We value your feedback CRM#” that contains a very quick survey to provide feedback on my support experience. I would really appreciate it if you can help me by filling it out as this feedback will help us improve the Microsoft App Assure experience.
It has been my pleasure to work with you.
Thank you,

**Comment by john-preston** [comment]
@Farce002 Thanks.

**Comment by Max-RM** [comment]
Can anyone solve this too?
https://github.com/telegramdesktop/tdesktop/issues/28757
